File: alex_hw/GenerateParanthesis.java
import java.io.*;
import java.util.*;
import java.lang.Math;

public class GenerateParanthesis {
    int count;

    public GenerateParanthesis() {
        count = 1;
    }

    public void traverse(String pattern, int target, int numLeft, int numRight) {
        if (numLeft + numRight > target) {
            return;
        }
        if (numLeft < numRight) {
            return;
        }
        if (numLeft + numRight == target) {
            if (numLeft == numRight) {
                System.out.printf("%d: %s\n", this.count, pattern);
                this.count++;
            }
        }
        this.traverse(pattern.concat("("), target, numLeft + 1, numRight);
        this.traverse(pattern.concat(")"), target, numLeft, numRight + 1);
    }

    public void generateParanthesis(int num) {
        if (num % 2 != 0) {
            System.out.println("Odd number of paranthesis!!!");
            return;
        }
        this.count = 1;
        System.out.printf("Number of parenthesis: %d \n", num);
        this.traverse("", num, 0, 0);
        System.out.println();
    }

    public static void main(String[] args) {
        GenerateParanthesis testCase = new GenerateParanthesis();
        testCase.generateParanthesis(2);
        testCase.generateParanthesis(4);
        testCase.generateParanthesis(6);
        testCase.generateParanthesis(8);
        testCase.generateParanthesis(10);
    }
}

File: alex_hw/generate_paranthesis.py
class GenerateParanthesis:
	def __init__(self):
		self.patterns = list()

	def traverse(self, pattern, target, num_left, num_right):
		if num_left + num_right > target:
			return

		if num_right > num_left:
			return

		if num_left + num_right == target:
			if num_left == num_right:
				self.patterns.append(pattern)
			return

		self.traverse(pattern + '(', target, num_left + 1, num_right)
		self.traverse(pattern + ')', target, num_left, num_right + 1)

	def generate_paranthesis(self, num):
		self.patterns = []
		if num % 2 != 0:
			print('Odd number of paranthesis!!!')

		self.traverse(pattern='', target=num, num_left=0, num_right=0)
		counter = 1
		print('number of paranthesis: %i' % num)
		for x in self.patterns:
			print('%i: %s' % (counter, x))
			counter += 1
		print('')


if __name__ == '__main__':
	test_case = GenerateParanthesis()
	test_case.generate_paranthesis(2)
	test_case.generate_paranthesis(4)
	test_case.generate_paranthesis(6)
	test_case.generate_paranthesis(8)
	test_case.generate_paranthesis(10)
